File: LeetCode/2392-successful-pairs-of-spells-and-potions/solution.cpp
class Solution {
public:
    // sort and binary search (lb)
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        int n=spells.size(), m=potions.size();
        sort(potions.begin(), potions.end());
        vector<int> res(n, 0);
        for (int i=0; i<n; i++) {
            int mid;
            int low = 0;
            int high = m;
            while (low < high) {
                mid = low + (high - low) / 2;
                if (success <= ((long long)spells[i])*potions[mid]) {
                    high = mid;
                }
                else {
                    low = mid + 1;
                }
            }

            if (low < m && ((long long)spells[i])*potions[low] < success) {
               low++;
            }

            
            int loc = low;
            if (loc == n) continue;
            res[i] = m - loc;
        }
        return res;
    }
};

File: LeetCode/2392-successful-pairs-of-spells-and-potions/solution.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        potions.sort()
        res = []
        for spell in spells:
            idx = bisect.bisect_left(potions, (success + spell - 1)//spell)
            res.append(len(potions) - idx)
        return res


